### ISSUE TITLE
Suggestion/Request/Contribution: Create a TNS service to access the Transient Name Server

### ISSUE THREAD

**Issue by emirkmo** [astropy/astroquery#2615]
TNS is the IAU approved catalog for all transient data.
Currently programmatic TNS access is locked behind a TNS account, creating a BOT key, and using the TNS API, and requires scientific credentials, none of which is good for open source or enabling access. (Although there is a public UI website to do all of this of course).

I propose that astroquery works with TNS to enable making TNS queries that follow the standard astroquery API and client-side caching. The problem is that TNS does not want to leak any API keys to the public, nor overload the searches. It is also not ideal to have to scrape the website, as breakages can happen and this can put strain on TNS servers.

The usecase I have in mind is two fold, one short term and one longer term and more involved:

1. **Short**: Query TNS for basic transient information such as name, ra & dec, type, redshift, discovery date. (Small json formatted payload of textual data). Pros: Low load, can be cached, query complexity can be tightly controlled. Cons: Do not get access to all TNS data.
2. **Long**: Query TNS for additional metadata for the transient, including things such as uploaded spectra. Search TNS for combinations of parameters. Pros: Do everything the TNS API can do, (and actually more). Cons: High load, easily abusable, cached results can be significantly wrong our outdated, query complexity can grow quickly. 

In a research group, I personally have a bot for doing TNS searches to obtain new transients, and have an API wrapper to accomplish the second set of goals. But this cannot scale to the whole public or easily be prevented from abuse.

However, number 1 IS easy. Currently, I self-host on Google Cloud the `tnsquery` API: https://tnsquery.darkdata.ml/api/docs
Which basically allows for doing limited size queries that return things like the name redshift, coordinates, etc., without leaking any API credentials, and allowing access to the public. See: https://github.com/emirkmo/tnsquery. The queries are cached locally at `tnsquery` API level in a performant and scalable DB so that repeat queries never even hit the TNS server, unless explicitly requested via a `force_tns = Ttue`  parameter.

I propose that astroquery uses a service like this, and we add a way to publicly query for basic transient information. In fact, it is totally fine to use `tnsquery` itself, until a longer term situation is figured out with TNS runners, as even querying the entire TNS db non-stop would not be a big cost, and it's possible to rate limit way before that. The cached DB and the rate limit of the `tnsquery` API key already secures 99% of what's needed.

Existing alternatives that for example use SIMBAD do not work as they do not have the latest transients. Nor can everyone use the TNS API since access requires scientific credentials. 

Additionally, there is a benefit that `Astropy`  itself can use this service to create `SkyCoord` objects from transient names, which is a hit or miss feature currently that requires the transient to be on Simbad or some other non-TNS source. I will open a related issue there.

**Comment by keflavich** [comment]
Thanks for the suggestion @emirkmo.   I think this would indeed be a valuable addition to astroquery, but you've highlighted some challenges.

Is https://www.wis-tns.org/ the formal host of the TNS?  It would be helpful to have one of the website hosters onboard to let us know what is/is not a good (allowed) use of their service.

Please do submit a pull request with your proposed module.

**Comment by emirkmo** [comment]
Yes it is. Great to hear that, let me add a PR and see how we can get someone from TNS on board. Would it work to message them with astropy dev mailing list CC'd?

**Comment by emirkmo** [comment]
Right, institutional support can be gotten if it will go into astroquery. :) I probably did not make that very clear. The name would be different. The service is a demo.

Good to hear about the scraping approach. We already do it as well but I will contact TNS because I wouldn’t want to be responsible for a DDOS on something many scientists rely on.

See the custom TNS search here for example: https://github.com/SNflows/tendrils/blob/da578e1eb56b9e75b3e0fd09d9de52fb13748994/tendrils/utils/tns.py#L129

Emir
________________________________
